### DISCUSSION TITLE
Can we not use UUID's as primary key in Keycloak database tables?

### DISCUSSION THREAD

**Discussion by jerrac** [keycloak/keycloak#32621] [Q&A] [answered]
I'm 99% sure the answer is "no", but figure it's worth asking anyway.

Can we configure Keycloak to use normal auto-increment row id's instead of UUID's stored in a varchar column?

UUID's as the row id column make inserts, and likely other, queries a LOT slower. In one application I developed, switching from UUID's to auto-incrementing ints, resulted in my performance speeding up by a LOT. At least 10x, but if I remember correctly it was even higher than that. Since I'm trying to push a bunch of users into Keycloak via the rest api, I could really use that extra performance. 

(I did search for related issues and discussions.)

**Comment by darius-m** [comment, marked as answer]
I'm sure there are also other reasons why the implementation uses UUIDs, but my guesses are that:
- auto incrementing indexes sometimes have collision issues when multiple instances try to create entries at the same time (and while UUIDs can sometimes generate collisions, they are far less common since the seed is random);
- using UUIDs can prevent account enumerations and other types of attacks, simply by making guessing user IDs harder (you could technically use an integer as the primary key, and have UUIDs as the user identifier for REST API requests, but since the UUIDs would still have to be unique indexes in the database, you would still have to check for their uniqueness when performing update operations);
- considering how many cryptographic operations an identity provider has to perform to validate requests, sign responses, I would assume that the increased load on the database by using UUIDs is negligible.

**Reply by jerrac** [reply]
Thanks. That does make sense. :)
